File: Solution Iterative.java

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Set;

public class Solution {

    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        if (rooms == null || rooms.size() <= 1) {
            return true;
        }

        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();
        queue.add(0);
        visited.add(0);

        while (!queue.isEmpty()) {
            int roomID = queue.poll();
            final List<Integer> keys = rooms.get(roomID);
            for (final int key : keys) {
                if (visited.add(key)) {
                    queue.add(key);
                }
            }
        }
        return visited.size() == rooms.size();
    }
}

File: Solution Iterative.cpp

#include <queue>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
  
public:
    bool canVisitAllRooms(vector<vector<int>>&rooms) {
        if (rooms.size() <= 1) {
            return true;
        }

        queue<int> queue;
        unordered_set<int> visited;
        visited.insert(0);
        queue.push(0);

        while (!queue.empty()) {
            const int roomID = queue.front();
            queue.pop();
            const vector<int>& keys = rooms[roomID];

            for (const auto& key : keys) {
                if (visited.find(key) == visited.end()) {
                    visited.insert(key);
                    queue.push(key);
                }
            }
        }
        return visited.size() == rooms.size();
    }
};
